### PR TITLE
UBC Driver Upgrade

### PR DESCRIPTION
**Pull request by gyrovorbis**
It dawned on me how useful the UBC could be for debugging, and how little of its power we have exposed publicly in our driver...

- Huge driver upgrade and new API which exposes everything it's actually capable of doing with an easy-to-use API that does the right thing and configures the most common use-cases with zero-initialization of the breakpoint structure.
- Bunch of new documentation explaining what the heck a UBC is, why it's useful, and how to configure the various types of breakpoints
- Created an example/test suite that demonstrates and flexes the various different kind of breakpoints that can be set with the UBC, ranging from simple instruction breakpoints to complex data breakpoints watching for ranges of values over ranges of addresses as well as sequential breakpoints.
- The UBC is used internally by the GDB stub; however, rather than using the UBC driver, it creates its own inner driver duplicating things and potentially interfering with outside usage. Phase 1 was to finish the new UBC driver, get it tested, and get it usable for `printf()`-style debugging in application code. Phase 2 will be to slide this under the `gdb_stub`, allowing us to implement more advanced breakpoint functionality for GDB.

### CODE REVIEW

File: examples/dreamcast/basic/breaking/breaking.c [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+   from a region of memory. */
+static bool break_on_sized_operand_region_access_value_range(void) {
+    uint32_t upper_boundary = 0;
+    alignas(1024) volatile uint32_t vars[1024 / sizeof(uint32_t)];
+    uint32_t lower_boundary = 0;
+
+    const ubc_breakpoint_t bp = {
+        .address = (void *)&vars,  // address to break on
+        .cond = {
+            .address_mask = ubc_address_mask_10, // don't care aabout lower 10 bits
```
Comment (andressbarajas) [inline examples/dreamcast/basic/breaking/breaking.c:156]:
```suggestion
            .address_mask = ubc_address_mask_10, // don't care about lower 10 bits
```

File: examples/dreamcast/basic/breaking/breaking.c [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    volatile uint16_t tmp;
+    tmp = var; (void)tmp;
+    VERIFY(!handled_); //we only did a read
+
+    var = 43;
+    VERIFY(!handled_); // we wrote the wrong value
+
+    *(uint8_t*)&var = 3;
+    VERIFY(!handled_); //we accessed it as the wrong size
```
Comment (andressbarajas) [inline examples/dreamcast/basic/breaking/breaking.c:135]:
```suggestion
    VERIFY(!handled_); // we accessed it as the wrong size
```

File: examples/dreamcast/basic/breaking/breaking.c [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    handled_ = false;
+
+    printf("Breaking on sized data write with value... ");
+
+    ubc_add_breakpoint(&bp, handlerfunc, (void *)true);
+
+    volatile uint16_t tmp;
+    tmp = var; (void)tmp;
+    VERIFY(!handled_); //we only did a read
```
Comment (andressbarajas) [inline examples/dreamcast/basic/breaking/breaking.c:129]:
```suggestion
    VERIFY(!handled_); // we only did a read
```

File: kernel/arch/dreamcast/include/dc/ubc.h [hunk 4]
```diff
@@ -1,113 +1,268 @@
...
+        You can only ever have a single sequential breakpoint active at a time,
+        and the maximum sequence length is 2 chained breakpoints. Data
+        comparison can only bet used in the second breakpoint of a sequence.
+    */
+    struct ubc_breakpoint *next;
+} ubc_breakpoint_t;
+
+/** \brief Breakpoint user callback
+
+    Typedef for the user function to be invoked upon a encountering a
```
Comment (andressbarajas) [inline kernel/arch/dreamcast/include/dc/ubc.h:199]:
```suggestion
    Typedef for the user function to be invoked upon encountering a
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    switch(addr_mask) {
+    case ubc_address_mask_all:
+        return 3;
+    case ubc_address_mask_16:
+        return 4;
+    case ubc_address_mask_20:
+        return 5;
+    default:
+        return addr_mask;
+    }
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:110]:
```suggestion
    switch(addr_mask) {
        case ubc_address_mask_all:
            return 3;
        case ubc_address_mask_16:
            return 4;
        case ubc_address_mask_20:
            return 5;
        default:
            return addr_mask;
    }
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+                         "nop\n");
+}
+
+/* Sets the state data and configures the UBC registers for the breakpoint on
+   the given channel. */
+static void enable_breakpoint(ubc_channel_t           ch,
+                              const ubc_breakpoint_t *bp,
+                              ubc_break_func_t        cb,
+                              void                   *ud) {
+
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:134]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    /* Wait for UBC configuration to refresh. */
+    ubc_wait();
+}
+
+/* Public entry-point for adding a breakpoint. */
+bool ubc_add_breakpoint(const ubc_breakpoint_t *bp,
+                        ubc_break_func_t       callback,
+                        void                   *user_data) {
+
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:201]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+void ubc_clear_breakpoints(void) {
+    disable_breakpoint(ubc_channel_b);
+    disable_breakpoint(ubc_channel_a);
+}
+
+/* Entry-point for UBC-related interrupt handling. */
+static void handle_exception(irq_t code, irq_context_t *irq_ctx) {
+    (void)code;
+
+    bool serviced = false;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:323]:
```suggestion
    bool serviced = false;

    (void)code;
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+    if(BRCR & CMFB) {
+        bool disable = false;
+
+        /* Invoke the user's callback if there is one. */
+        if(channel_state[ubc_channel_b].cb)
+            disable = channel_state[ubc_channel_b].cb(
+                            channel_state[ubc_channel_b].bp,
+                            irq_ctx,
+                            channel_state[ubc_channel_b].ud
+                        );
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:335]:
```suggestion
            disable = channel_state[ubc_channel_b].cb(
                            channel_state[ubc_channel_b].bp,
                            irq_ctx,
                            channel_state[ubc_channel_b].ud);
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+            if(BRCR & SEQ)
+                disable_breakpoint(ubc_channel_a);
+        }
+
+        serviced = true;
+    }
+
+    /* Check if channel A's condition is active. */
+    if(BRCR & CMFA) {
+
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:351]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/ubc.c [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+        if(!(BRCR & SEQ)) {
+            bool disable = false;
+
+            /* Invoke the user's callback if there is one. */
+            if(channel_state[ubc_channel_a].cb)
+                disable = channel_state[ubc_channel_a].cb(
+                                channel_state[ubc_channel_a].bp,
+                                irq_ctx,
+                                channel_state[ubc_channel_a].ud
+                            );
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/ubc.c:363]:
```suggestion
                disable = channel_state[ubc_channel_a].cb(
                                channel_state[ubc_channel_a].bp,
                                irq_ctx,
                                channel_state[ubc_channel_a].ud);
```

**Review by ljsebald — CHANGES_REQUESTED** [review]

File: examples/dreamcast/basic/breaking/breaking.c [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+    success &= break_on_sequence();
+
+    if(success) {
+        printf("\n***** Breakpoint Test: SUCCESS *****\n");
+        return EXIT_SUCCESS;
+    } else {
+        fprintf(stderr, "\n***** Breakpoint Test: FAILURE *****\n");
+        return EXIT_FAILURE;
+    }
+}
```
Comment (ljsebald) [inline examples/dreamcast/basic/breaking/breaking.c:358]:
```suggestion
}

```

**Comment by gyrovorbis** [comment]
Retested after addressing review feedback + resolving conflicts, and everything is still working great!

**Review by ljsebald — APPROVED** [review]

**Pull request merged**